ui/raidboss/emulator/ui/EncounterTab.ts

```diff
@@ -0,0 +1,285 @@
+      }
+
+      this.refreshUI();
+    });
+  }
+
+  refreshUI(): void {
+    this.refreshZones();
+    this.refreshDates();
+    this.refreshEncounters();
+    this.refreshInfo();
+  }
+
+  refreshZones(): void {
+    this.$zoneColumn.innerHTML = '';
+
+    let clear = true;
+
+    const zones = new Set(Object.keys(this.encounters));
+
+    for (const zone of [...zones].sort()) {
+      const $row = this.$encounterTabRowTemplate.cloneNode(true);
+      if (!($row instanceof HTMLElement))
+        throw new UnreachableCode();
+      $row.innerText = zone;
+      if (zone === this.currentZone) {
+        clear = false;
+        $row.classList.add('selected');
+      }
+      $row.addEventListener('click', (ev) => {
+        const t = ev.currentTarget;
+        if (!(t instanceof HTMLElement))
+          throw new UnreachableCode();
+        const parent = t.parentElement;
+        if (!parent)
+          throw new UnreachableCode();
+        querySelectorAllSafe(parent, '.selectorRow.selected').forEach((n) => {
+          n.classList.remove('selected');
+        });
+        t.classList.add('selected');
+        this.currentZone = t.textContent ?? undefined;
+        this.refreshUI();
+      });
+      this.$zoneColumn.append($row);
+    }
+
+    if (clear)
+      this.currentZone = undefined;
+  }
+
+  refreshDates(): void {
+    this.$dateColumn.innerHTML = '';
+
+    let clear = true;
+
+    if (this.currentZone !== undefined) {
+      const zoneMap = this.encounters[this.currentZone];
+      if (!zoneMap)
+        return;
+      const dates = new Set<string>(Object.keys(zoneMap));
+      for (const date of [...dates].sort()) {
+        const $row = this.$encounterTabRowTemplate.cloneNode(true);
+        if (!($row instanceof HTMLElement))
+          throw new UnreachableCode();
+        $row.innerText = date;
+        if (date === this.currentDate) {
+          clear = false;
+          $row.classList.add('selected');
+        }
+        $row.addEventListener('click', (ev) => {
+          const t = ev.currentTarget;
+          if (!(t instanceof HTMLElement))
+            throw new UnreachableCode();
+          const parent = t.parentElement;
+          if (!parent)
+            throw new UnreachableCode();
+          querySelectorAllSafe(parent, '.selectorRow.selected').forEach((n) => {
+            n.classList.remove('selected');
+          });
+          t.classList.add('selected');
+          this.currentDate = t.textContent ?? undefined;
+          this.refreshUI();
+        });
+        this.$dateColumn.append($row);
+      }
+    }
+
+    if (clear)
+      this.currentDate = undefined;
+  }
+
+  refreshEncounters(): void {
+    this.$encounterColumn.innerHTML = '';
+
+    let clear = true;
+
+    const zoneMap = this.currentZone ? this.encounters[this.currentZone] : undefined;
+
+    if (!zoneMap)
+      return;
+
+    const dateMap = this.currentDate ? zoneMap[this.currentDate] : undefined;
+
+    if (!dateMap)
+      return;
```

Comment:
```suggestion
    if (!this.currentZone || !this.currentDate)
      return;

    const zoneMap = this.encounters[this.currentZone];
    if (!zoneMap)
      return;
  
    const dateMap = zoneMap[this.currentDate];
    if (!dateMap)
      return;
```

What about something more like this to avoid the unnecessary ternary?

Reply:
This was the first of these three that I converted and I think it's a bit obvious I improved the code after this one 😉